pkg/graph/resolvers/mto_milestone.go

```diff
@@ -78,7 +85,50 @@ func MTOMilestoneCreateCommon(ctx context.Context, logger *zap.Logger, principal
 		}
 
 		// A common milestone never has a name (since it comes from the Common Milestone itself), so pass in `nil`
-		milestone := models.NewMTOMilestone(principalAccount.ID, nil, &commonMilestoneKey, modelPlanID, &finalCategoryID)
+		milestone := models.NewMTOMilestone(
+			principalAccount.ID,
+			nil,
+			&commonMilestoneKey,
+			modelPlanID,
+			&finalCategoryID,
+		)
+
+		keys := lo.Map(commonSolutions, func(commonSolutionKey models.MTOCommonSolutionKey, _ int) storage.MTOSolutionGetByModelPlanIDAndCommonSolutionKey {
+			return storage.MTOSolutionGetByModelPlanIDAndCommonSolutionKey{
+				ModelPlanID:          modelPlanID,
+				MTOCommonSolutionKey: commonSolutionKey,
+			}
+		})
+
+		solutions, err := storage.MTOSolutionGetByModelPlanIDAndCommonSolutionKeyLoader(tx, logger, keys)
+
+		for i, solution := range solutions {
+			if solution == nil {
+				// TODO: Can this be moved upstream as part of the loader?
+				// TODO: Is there another more proper way to do this?
+				// TODO: Are non-existent solutions returned as nil? Can they be?
+				solution, err = MTOSolutionCreateCustom(
```

Comment:
@StevenWadeOddball This relates to another comment in the query. Thinking about the proper query-logic interface design here. I think it's about 90% there, but obviously some nuance remains.

Reply:
Thanks Tom,

The other approach we briefly discussed was just trying to do an insert. So you could follow the insert common solution logic, but you just provide a list of common solution keys. If their is a record, you don't insert. If there isn't one, you insert.

Then you select and return all the solutions. With that approach, you don' have to iterate through every possible solution, or figure out if it exists or not, you just try to make all of them, and get a list of all of them. I think you can do the same thing with the links afterwards as well, just try to create them, and just ignore the insert attempt if the link already exists.  (another case for ON CONFLICT, or potentially a merge statement)

How do you think you want to proceed?

Reply:
Thanks for the input Steven, between your comment and a chat with Clay and after carefully weighing the options I agree that an upsert approach is better here and updated the logic